### ISSUE TITLE
Update GET /collections/{collection_id}/datasets/{dataset_id} to include all legacy schema metadata

### ISSUE THREAD

**Issue by brianraymor** [chanzuckerberg/single-cell#420]
---

Capturing a private conversation with @jahilton that requires further assessment and refinement :_

> ... Wanted to see how feasible these would be to add to the /collections/{collection_id}/datasets/{dataset_id} endpoint...

1. feature count (var length)
2. feature_biotype (all values present)
3. feature_reference (all values present)
4. default_embedding (if defined in uns)
5. ~layers present~ (.X + raw.X if present + ~layers keys~)
6. embeddings present (obsm keys)

@brianraymor notes:

I struck **layers** because the schema does not reference/require [anndata layers](https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.layers.html). Currently, the Discover API only returns fields defined in the schema.

@nayib-jose-gloria notes:

For assigned engineer--add fields in DatasetMetadata entity object, but make them Optional/nullable to support hybrid migration state between versions. Populate in dataset_metadata JSON blob DatasetVersion DB table in extract_metadata. For fields that say 'all values present', extract + assign the set of field values in dataset.

Backfill will occur by default during Schema 4.0.0 migration.

**Comment by nayib-jose-gloria** [comment]
@brianraymor Just to confirm, does "all values present" translate to a list w/o regard to a particular ordering that includes all unique values for that field in the dataset?

**Comment by brianraymor** [comment]
@jahilton - my interpretation of "all values present" was the **set** of values rather than the list. Can you confirm or deny?

**Comment by jahilton** [comment]
confirmed **set** - i.e. all unique values (no particular ordering)

**Comment by nayib-jose-gloria** [comment]
^ this was also how I parsed it, but I decided to try to describe a set instead of saying set for some reason LOL

**Comment by nayib-jose-gloria** [comment]
@brianraymor @jahilton should these appear in the dataset index endpoint as well? or just the single dataset GET endpoint? Further, should we extend this to the data portal API?

**Comment by brianraymor** [comment]
> should these appear in the dataset index endpoint as well?

These are primarily intended for curators / data generation, so I think not. @pablo-gar - do you see any useful scenarios for census? 

> Further, should we extend this to the data portal API?

There's no rationale to include in the DP API because the information is not being surfaced in the UX.

**Comment by nayib-jose-gloria** [comment]
@brianraymor up to this point, the dataset index has returned the same metadata as the single dataset GET endpoint so I am tempted to keep it that way and add these fields there as well if we have the flexibility to do so. But of course, we can deviate if we prefer to keep it smaller

**Comment by pablo-gar** [comment]
> These are primarily intended for curators / data generation, so I think not. @pablo-gar - do you see any useful scenarios for census?

I'll respond for each:

> feature count (var length)

Census does not use this

> feature_biotype (all values present)

Census does not use this, but it _could_ be useful when expanding to non-RNA data. Nor current plans of doing so.

> feature_reference (all values present)

Census uses this but currently gets it from AnnData and that is sufficient.

> default_embedding (if defined in uns)

Census does not use this

> embeddings present (obsm keys)

Census does not use this

**Comment by nayib-jose-gloria** [comment]
@brianraymor 

Proposed [swagger](https://pr-5891-backend.rdev.single-cell.czi.technology/curation/ui/) for change. 

Diff: Includes new dataset metadata fields in single dataset GET, single collection GET with full dataset metadata, single dataset_version GET, and dataset/{dataset_id}/versions GET. also introduced descriptions for each, please review. 

PR view is https://github.com/chanzuckerberg/single-cell-data-portal/pull/5891/files#diff-d808b5972f311fbcd931698c57088028785292368fca7b9b8499a9da60578c9b

**Comment by brianraymor** [comment]
@jahilton - would you also like to review since this is your request? 

I think these are the direct links:

[single dataset GET](https://pr-5891-backend.rdev.single-cell.czi.technology/curation/ui/#/Collection/backend.curation.api.v1.curation.collections.collection_id.datasets.dataset_id.actions.get)

[single collection GET with full dataset metadata](https://pr-5891-backend.rdev.single-cell.czi.technology/curation/ui/#/Collection/backend.curation.api.v1.curation.collections.collection_id.actions.get)

[single dataset_version GET](https://pr-5891-backend.rdev.single-cell.czi.technology/curation/ui/#/Dataset%20Version/backend.curation.api.v1.curation.dataset_versions.dataset_version_id.actions.get)

[dataset/{dataset_id}/versions GET](https://pr-5891-backend.rdev.single-cell.czi.technology/curation/ui/#/Dataset/backend.curation.api.v1.curation.datasets.dataset_id.versions.actions.get)

**Comment by brianraymor** [comment]
I don't see:

5. ~layers present~ (.X + raw.X if present + ~layers keys~)

Layers was struck because the schema does not reference layers per my note, but .X and .raw.X were still required. 

---

Otherwise, only an editorial nit. I would reference embedding **keys** in the descriptions such as:

Key for the embedding that is displayed by default in CZ CELLxGENE Explorer

List of keys for embeddings in the dataset OR List of embedding keys in the dataset

**Comment by jahilton** [comment]
LGTM, except the .X/raw.X as noted

**Comment by nayib-jose-gloria** [comment]
@brianraymor @jahilton Sorry, I assumed the strikethrough was just formatted incorrectly. Can I get more details on what you want to return for '.X and raw.X if present'? Is it a list that is either empty, ['X'], or ['X', 'raw.X']? Two flag fields for X and raw.X like has_X and has_raw_X? The former is more extensible should we decide we want to support other layers being listed in the future. 

another point of confusion--I thought the X matrix was required. Does this imply that its not? Or that I'm totally misunderstanding the ask and there's some other metadata you want to return in the API response?

**Comment by jahilton** [comment]
tldr i haven't thought of the ideal solution

What I'd like to get from the API is whether the RNA datasets have optionally submitted a normalized data layer.
Another way to ask that - is raw empty in the downloaded AnnData or not?
Yet another way - where are the raw counts located in the downloaded AnnData?

You are correct, X is required by AnnData.

Options that would satisfy the ask:

- `has_raw_X` boolean (a little concerned that someone seeing False here would interpret that as "there is no raw data")
- `raw_X_min`, `raw_X_max`, `X_min`, `X_max` if the min/max values of each layer are readily available. raw_X_* are not present if raw is empty. The range could be useful for other reasons.
- `raw_location` either `X` or `raw.X` for RNA. non-RNA is `X` unless raw is not empty (some submit "raw" data optionally)

Those (or any other idea I can come up with) are AnnData-centric, but maybe there's no getting around that.

**Comment by nayib-jose-gloria** [comment]
@jahilton For an option revolving less around Anndata terms, what do you think of `has_normalized_data` where, for RNA assays, True means `raw.X` exists and False means it does not. For non-RNA assays, it is always True.

**Comment by nayib-jose-gloria** [comment]
> @jahilton For an option revolving less around Anndata terms, what do you think of `has_normalized_data` where, for RNA assays, True means `raw.X` exists and False means it does not. For non-RNA assays, it is always True.

actually, unless you strongly prefer this, I will stick with the `raw_location` approach you suggested, its simpler on the backend

**Comment by jahilton** [comment]
@nayib-jose-gloria `raw_location` sounds good

**Comment by nayib-jose-gloria** [comment]
@jahilton @brianraymor 

FYI I ended up using 'raw_data_location' as the field name to be a bit more specific. Can easily switch it back if desired. 

Also, I did not hide this behind a feature flag because, technically, we can start populating these fields for new uploads as soon as we merge since everything is already on the schema 3 h5ads. However, it would be pretty involved + costly to try to backfill populate previously uploaded datasets including old versions. Once we migrate, all the current public datasets will have these fields backfilled. Old versions however will still not have the fields backfilled in the API. 

For the immediate state: do we want to hide this behind a feature flag and populate everything at once in the API after schema 4 migration? Or are we ok with new uploads picking this up now in the API and currently published datasets either returning null for these fields (or simply not return these fields) in the interim until migration? 

For old versions: I would prefer not doing a backfill as it would greatly increase the cost+scope of the issue, but let me know if you disagree. If we do not backfill, old versions will return these fields as null (or simply not return these fields).

**Comment by brianraymor** [comment]
> For old versions: I would prefer not doing a backfill as it would greatly increase the cost+scope of the issue, but let me know if you disagree. If we do not backfill, old versions will return these fields as null (or simply not return these fields).

I agree that there's no reason to backfill old versions in the interim.

I don't see a strong need for a feature flag in this case. It has zero impact on current consumers. 

In general, are you introducing other API changes behind a feature flag? For example, all the changes related to `tissue_type` which do impact existing consumers.

**Comment by jahilton** [comment]
No concerns from me on implementing going forward without backfilling until migration.

**Comment by nayib-jose-gloria** [comment]
> In general, are you introducing other API changes behind a feature flag? For example, all the changes related to `tissue_type` which do impact existing consumers.

yup!
